examples/asrc_demo/src/usb/usb_audio.c

```diff
@@ -0,0 +1,1084 @@
+{
+    (void)rhport;
+
+    // Page 91 in UAC2 specification
+    uint8_t channelNum = TU_U16_LOW(p_request->wValue);
+    uint8_t ctrlSel = TU_U16_HIGH(p_request->wValue);
+    uint8_t itf = TU_U16_LOW(p_request->wIndex);
+
+    (void)channelNum;
+    (void)ctrlSel;
+    (void)itf;
+
+    return false; // Yet not implemented
+}
+
+// Invoked when audio class specific get request received for an entity
+bool tud_audio_get_req_entity_cb(uint8_t rhport,
+                                 tusb_control_request_t const *p_request)
+{
+    (void)rhport;
+
+    // Page 91 in UAC2 specification
+    uint8_t channelNum = TU_U16_LOW(p_request->wValue);
+    uint8_t ctrlSel = TU_U16_HIGH(p_request->wValue);
+    // uint8_t itf = TU_U16_LOW(p_request->wIndex); 			// Since we have only one audio function implemented, we do not need the itf value
+    uint8_t entityID = TU_U16_HIGH(p_request->wIndex);
+
+    // Input terminal (Microphone input)
+    if (entityID == UAC2_ENTITY_MIC_INPUT_TERMINAL)
+    {
+        switch (ctrlSel)
+        {
+        case AUDIO_TE_CTRL_CONNECTOR:;
+            // The terminal connector control only has a get request with only the CUR attribute.
+
+            audio_desc_channel_cluster_t ret;
+
+            // Those are dummy values for now
+            ret.bNrChannels = CFG_TUD_AUDIO_FUNC_1_N_CHANNELS_TX;
+            ret.bmChannelConfig = 0;
+            ret.iChannelNames = 0;
+
+            TU_LOG2("    Get terminal connector\r\n");
+            rtos_printf("Get terminal connector\r\n");
+
+            return tud_audio_buffer_and_schedule_control_xfer(rhport, p_request, (void *)&ret, sizeof(ret));
+
+            // Unknown/Unsupported control selector
+        default:
+            TU_BREAKPOINT();
+            return false;
+        }
+    }
+
+    // Feature unit
+    if (entityID == UAC2_ENTITY_MIC_FEATURE_UNIT)
+    {
+        switch (ctrlSel)
+        {
+        case AUDIO_FU_CTRL_MUTE:
+            // Audio control mute cur parameter block consists of only one byte - we thus can send it right away
+            // There does not exist a range parameter block for mute
+            TU_LOG2("    Get Mute of channel: %u\r\n", channelNum);
+            return tud_control_xfer(rhport, p_request, &mute[channelNum], 1);
+
+        case AUDIO_FU_CTRL_VOLUME:
+
+            switch (p_request->bRequest)
+            {
+            case AUDIO_CS_REQ_CUR:
+                TU_LOG2("    Get Volume of channel: %u\r\n", channelNum);
+                return tud_control_xfer(rhport, p_request, &volume[channelNum], sizeof(volume[channelNum]));
+            case AUDIO_CS_REQ_RANGE:
+                TU_LOG2("    Get Volume range of channel: %u\r\n", channelNum);
+
+                // Copy values - only for testing - better is version below
+                audio_control_range_2_n_t(1) ret;
+
+                ret.wNumSubRanges = 1;
+                ret.subrange[0].bMin = -90; // -90 dB
```

Comment:
I think these values didn't work? Better to copy from XVF3800 if volume control should work

Reply:
Volume control is not supported in the descriptors so this never gets used. I'll remove this code. Will worry about this when/if I add vol. control.

Reply:
I've added the vol. defines from xvf3820 and added a TODO about missing vol. control implementation.